src/scripts/consume_logs/consume_logs.py

```diff
@@ -0,0 +1,312 @@
+    )
+    _check_for_failed_query(query_response)
+    while query_response["status"] != "Complete":
+        time.sleep(1)
+        query_response = logs_client.get_query_results(
+                queryId=start_query_response["queryId"]
+        )
+        _check_for_failed_query(query_response)
+    return query_response["results"]
+
+def _check_for_failed_query(query_response):
+    """
+    Helper function for checking whether a query response has not succeeded.
+    """
+    if query_response["status"] in ["Failed", "Cancelled", "Timeout", "Unknown"]:
+        raise UserWarning(f"Query failed with status \"{query_response['status']}\"")
+
+def group_query_result_by_workflow_run(query_results: List[List[dict]]) -> Dict[str, List[dict]]:
+    """
+    Associates log records with their workflow run ID.
+
+    Since our log query may have picked up logs from multiple workflow runs,
+    we organize our log records by mapping the workflow run ID to its respective
+    log records.
+
+    Args:
+        query_results (list[list[dict]]): The query results. See `get_query_results`.
+
+    Returns:
+       (dict[str, list[dict]]): A dict of workflow run ID mapped to its respective logs.
+    """
+    # e.g., [{'@message': '{...}', '@ptr': '...'}, ...]
+    log_records = [
+            {field['field']: field['value'] for field in log_message}
+            for log_message in query_results
+    ]
+    workflow_run_logs = defaultdict(list)
+    for log_record in log_records:
+        log_message = json.loads(log_record["@message"])
+        workflow_run_logs[log_message["process"]["parent"]["pid"]].append(log_message)
+    return workflow_run_logs
+
+def transform_logs_to_dataframe(log_messages: List[dict]) -> "pandas.DataFrame":
+    """
+    Construct a pandas DataFrame from log records
+
+    Dataframes will contain information necessary for comparing read/write
+    line counts of individual files from a workflow run.
+
+    Args:
+        log_messages (list[dict]): A list of log messages
+
+    Returns:
+        pandas.DataFrame: A pandas DataFrame with columns:
+            * cohort (str)
+            * file_name (str)
+            * event_type (str, either 'access' or 'creation')
+            * line_count (int)
+    """
+    dataframe_records = []
+    for log_message in log_messages:
+        if (
+                "event" not in log_message
+                or "type" not in log_message["event"]
+                or not any(
+                    [
+                        k in log_message["event"]["type"]
+                        for k in ["access", "creation"]
+                    ]
+                )
+                or all(
+                    [
+                        k in log_message["event"]["type"]
+                        for k in ["access", "creation"]
+                    ]
+                )
+        ):
+            raise KeyError(
+                    "Did not find event.type in log message or "
```

Comment:
What would cause this?

Reply:
In general, what will break the assumptions of this script is if we added/updated logs which match the query (`file.action` = `list-file-properties`) but don't conform to the format we expect from such logs. This particular case checks that `event.type` contains "access" xor "creation". This would never happen with the current implementation of S3 to JSON, but changes we make in the future could potentially break the assumptions we make if we aren't careful.